handout/testing/Introduction to TDD Testing.md

```diff
@@ -0,0 +1,192 @@
+# Introduction to TDD Testing
+
+What is Test-Driven-Development? Well, simply put, its an engineering process in which the developer writes an intial automated test case that defines a feature, then writes the minimum amount of code to pass the test and eventually refactors the code to acceptable standards. 
+
+A **unit test** is used to test individual components of the system. An **integration test** is a test which tests the system as a whole, and how it will run in production.
+
+Unit tests should only verify the behaviour of a specific unit of code. If the unit's behaviour is modified, then the unit test would be updated as well. Unit tests should not make assumptions about the behaviour of *other* parts of your codebase or your dependencies. When other parts of your codebase are modified, your unit tests **should not fail**. (If they do fail, you have written a test that relies on other components, it is therefore not a unit test.) Unit tests are cheap to maintain, and should only be updated when the individual units are modified. For TDD in Angular, a unit is most commonly defined as a class, pipe, component, or service. It is important to keep units relatively small, this keeps tests in a 'self-documenting' state, where they are easy to read and understand. 
+
+## Toolchain
+
+Our testing toolchain will consist of the following tools. 
+
+- Jasmine
+- Karma
+- Phantom-js
+- Istanbul
+- Sinon
+- chai
+
+[Jasmine](http://jasmine.github.io/) is the testing framework that has been baked into the Angular 2 `angular2/testing` module. This is the core framework that we will write our unit tests with. 
+
+[Karma](https://karma-runner.github.io/) is a test automation tool for controlling the execution of our tests, what platform to perform them under, and allows us to generate various reports on the results. For one or two tests this may seem like overkill, but as an application grows larger and the number of units to test grows, it is important to organize, execute, and report on tests in an efficient manner. Karma is library agnostic so we could use other testing frameworks in combination with other tools (like code coverage reports, spy testing, e2e, etc.). 
+
+In order to test our Angular 2 application we need create an environment for it to run in. We could use a browser like Chrome or Firefox to accomplish this (Karma supports in-browser testing). We could also use a browser-less environment to test our application, which can offer us greater control over automating certain tasks and managing our testing workflow. [Phantom-js](http://phantomjs.org/) provides a javascript api that allows us to create a headless DOM instance which can be used to bootstrap our Angular 2 application. Then, using that DOM instance that is running our Angular 2 application, we can run our tests. 
```

Comment:
"In order to test our Angular 2 application we to create an environment for it to run."

"provides a Javascript API that allows us..."